taskcat/cfn/template.py

```diff
@@ -0,0 +1,133 @@
+import cfnlint
+from pathlib import Path
+from taskcat.client_factory import ClientFactory
+from taskcat.common_utils import s3_url_maker
+from typing import List, Set
+import logging
+import random
+import string
+
+LOG = logging.getLogger(__name__)
+
+
+class Template:
+
+    def __init__(self, template_path: str, project_root: str = '', url: str = '',
+                 client_factory_instance: ClientFactory = ClientFactory()):
+        self.template_path: Path = Path(template_path).absolute()
+        self.template = cfnlint.decode.cfn_yaml.load(self.template_path)
+        with open(template_path, 'r') as fh:
+            self.raw_template = fh.read()
+        project_root = project_root if project_root else self.template_path.parent.parent
+        self.project_root = Path(project_root).absolute()
+        self.client_factory_instance = client_factory_instance
+        self.url = url
+        self.children: List[Template] = []
+        self._find_children()
+
+    def __str__(self):
+        return str(self.template)
+
+    def __repr__(self):
+        return "<Template {} at {}>".format(self.template_path, hex(id(self)))
+
+    def _upload(self, bucket_name: str, prefix: str = '') -> str:
+        s3_client = self.client_factory_instance.get('s3')
+        s3_client.upload_file(str(self.template_path), bucket_name, prefix + self.template_path.name)
+        return s3_url_maker(bucket_name, f'{prefix}{self.template_path.name}', self.client_factory_instance.get('s3'))
+
+    def _delete_s3_object(self, url):
+        if url:
+            bucket_name = url.split('//')[1].split('.')[0]
+            path = '/'.join(url.split('//')[1].split('/')[1:])
+            s3_client = self.client_factory_instance.get('s3')
+            s3_client.delete_objects(Bucket=bucket_name, Delete={'Objects': [{'Key': path}], 'Quiet': True})
+
+    def write(self):
+        """writes raw_template back to file, and reloads decoded template, useful if the template has been modified"""
+        with open(str(self.template_path), 'w') as fh:
+            fh.write(self.raw_template)
+        self.template = cfnlint.decode.cfn_yaml.load(self.template_path)
+        self._find_children()
+
+    def validate(self, region, bucket_name: str = '', prefix: str = ''):
+        cfn_client = self.client_factory_instance.get('cloudformation', region)
+        if not self.url and not bucket_name:
+            raise ValueError("validate requires either the url instance variable, or bucket_name+prefix to be provided")
+        tmpurl = ''
+        if not self.url:
+            rand = ''.join(random.choice(string.ascii_lowercase) for i in range(8)) + '/'
+            tmpurl = self._upload(bucket_name, prefix+rand)
```

Comment:
Help me understand why we're uploading to S3 when we can validate locally?

Reply:
`TemplateBody` has a much smaller size limit

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudformation.html#CloudFormation.Client.validate_template

> TemplateURL (string) -- max size: 460,800 bytes
> TemplateBody (string) -- maximum length of 51,200 bytes.